### PR TITLE
[SPARK-14189][SQL] JSON data sources find compatible types even if inferred decimal type is not capable of the others

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

https://issues.apache.org/jira/browse/SPARK-14189

When inferred types in the same field during finding compatible `DataType`, are `IntegralType` and `DecimalType` but `DecimalType` is not capable of the given `IntegralType`, JSON data source simply fails to find a compatible type resulting in `StringType`.

This can be observed when `prefersDecimal` is enabled.

``` scala
def mixedIntegerAndDoubleRecords: RDD[String] =
  sqlContext.sparkContext.parallelize(
    """{"a": 3, "b": 1.1}""" ::
    """{"a": 3.1, "b": 1}""" :: Nil)

val jsonDF = sqlContext.read
  .option("prefersDecimal", "true")
  .json(mixedIntegerAndDoubleRecords)
  .printSchema()
```
- **Before**

```
root
 |-- a: string (nullable = true)
 |-- b: string (nullable = true)
```
- **After**

```
root
 |-- a: decimal(21, 1) (nullable = true)
 |-- b: decimal(21, 1) (nullable = true)
```

(Note that integer is inferred as `LongType` which becomes `DecimalType(20, 0)`)
## How was this patch tested?

unit tests were used and style tests by `dev/run_tests`.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @rxin

**Comment by HyukjinKwon** [comment]
@davies Could you maybe take a look at this one please? It is similar with https://github.com/apache/spark/pull/12030.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/TestJsonData.scala [hunk 1]
```diff
@@ -214,6 +214,11 @@ private[json] trait TestJsonData {
       """{"a": {"b": 1}}""" ::
       """{"a": []}""" :: Nil)
 
+  def mixedIntegerAndDoubleRecords: RDD[String] =
```
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/TestJsonData.scala]: This is only used in one test case, it's better to put inside the test case.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/InferSchema.scala [hunk 2]
```diff
@@ -256,6 +256,14 @@ private[sql] object InferSchema {
         case (ArrayType(elementType1, containsNull1), ArrayType(elementType2, containsNull2)) =>
           ArrayType(compatibleType(elementType1, elementType2), containsNull1 || containsNull2)
 
+        // The case tha given `DecimalType` is capable of given `IntegralType` is handled in
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/InferSchema.scala]: tha ?

**Comment by davies** [comment]
LGTM, only some minor comments.

**Comment by HyukjinKwon** [comment]
@rxin Could you maybe take another look or merge this please?

**Comment by rxin** [comment]
Thanks - merging in master.

**Pull request closed without merging**